kythe/typescript/post_processing.ts

```diff
@@ -0,0 +1,140 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Post processing that takes data produced by the indexer and
+ * modifies it to handle cases which can't be effectively or easily handled
+ * during normal indexing. Code here works only with Kythe edges and facts and
+ * doesn't use TS AST.
+ */
+
+
+import {EdgeKind, JSONEdge, KytheData, VName} from './kythe';
+
+/**
+ * Convert VName to a string that can be used as key in Maps.
+ */
+function vnameToString(vname: VName): string {
+    return `(${vname.corpus},${vname.language},${vname.path},${vname.root},${
+        vname.signature})`;
+  }
+
+/**
+ * Finds node pairs that represent imported nodes. Key - string representation of a vname of the
+ * local node e.g. foo in `import {foo} from './bar';` and value is VName of the original node
+ * e.g. `foo` in `export const foo = 1;`.
+ *
+ * All edges that point to the first node (local) will be reassigned to second (original).
+ */
+function findImportNodesThatNeedEdgesReassignment(data: Readonly<KytheData>): Map<string, VName> {
+    const result = new Map<string, VName>();
+    // Map of anchor ==ref/import==> node.
+    const importRefs = new Map<string, VName>();
+    // Map of anchor ==defines/binding==> node.
+    const defines = new Map<string, VName>();
+    for (const entry of data) {
+      if (!('target' in entry)) continue;
+      if (entry.edge_kind === EdgeKind.DEFINES_BINDING) {
+        defines.set(vnameToString(entry.source), entry.target);
+      }
+      if (entry.edge_kind === EdgeKind.REF_IMPORTS) {
+        importRefs.set(vnameToString(entry.source), entry.target);
+      }
+    }
+    for (const [anchor, node] of importRefs.entries()) {
+      const define = defines.get(anchor);
+      if (define != null) {
+        result.set(vnameToString(define), node);
+      }
+    }
+    return result;
+
+}
+
+/**
+ * This method merges nodes for imported symbols.
+ *
+ * Indexer produces the following graph:
+ *
+ * // foo.ts
+ * //- @ANSWER defines/binding AnswerOrig
+ * export const ANSWER = 42;
+ *
+ * // bar.ts
+ * //- @ANSWER ref/imports AnswerOrig
+ * //- @ANSWER defines/binding AnswerLocal
+ * import {ANSWER} from './foo';
+ * //- @ANSWER ref AnswerLocal
+ * console.log(ANSWER);
+ *
+ *
+ * This pass `reassigneEdgesForImports` changes
+ *
+ * //- @ANSWER ref AnswerLocal
+ * console.log(ANSWER);
+ *
+ * to
+ *
+ * //- @ANSWER ref AnswerOrig
+ * console.log(ANSWER);
+ *
+ * and removes `@ANSWER defines/binding AnswerLocal` definition as it doesn't
+ * have any refs to it now.
+ *
+ *
+ * Notice that ANSWER anchor in console.log call now points at the AnswerOrig node.
+ * This is what users expect: that usages of imported symbols will point to the
+ * original definition and not local aliases.
+ */
+function reassignEdgesForImports(data: Readonly<KytheData>): KytheData {
+    const result: KytheData = [];
```

Comment:
Maybe premature optimization, but we've seen memory issues in the TS indexer.

So we're not just buffering all the emitted data, but buffering it 2x here. Is there a way to avoid this? Can we at least drop references to the AST and file data before doing this? Is there a way to do this my mutating the original data in-place?

Reply:
Actually it wouldn't be 2x here. We create a copy of an array but we are not copying entries themselves. We mutating them in-place. `Readonly<KytheData>` is a bit misleading: it means we can't modify array itself but can and do modify elements of the array.

As for dropping AST or file data - I don't think it can be controlled in TS/JS. I believe that NodeJS should be able to clear those objects as they are unused by that point. I added enablePostProcessing flag so it should help us to see impact of just adding buffering. 

Mutating array fully in-place a bit clunky as we need to remove some entries from the array. So we would need to make array elements nullable and then carefully handle nulls. Which is a bit annoying. But if it turns out that enabling post processing has bad impact - I'll do that.

Reply:
For historical context, this specific issue (memory overload) was an important motivation for some of Kythe's design. We had Python indexer shards falling over because they needed to retain the whole indexer output for a module in-memory, and since the dependencies are all source or source-equivalent, it didn't fit in some cases.

As a rule, this is why we designed the graph schema not to do denormalizations "directly" in the indexer, but instead to handle them as postprocessing steps. It's safer in languages with separate compilation, but TS is not one of those 🙂.

One can't necessarily avoid having to have all the source in-memory, but ideally the indexer should try hard to not keep the graph artifacts in-memory. In some cases one can get away with a cache of smaller pieces, though.

Reply:
Actually I probably can do that inlining without buffering, by complicating TS indexer itself.

Reply:
Rewrote implementation by having local per-source file vname => vname mapping. Similar like we do in JS indexer.